scan-command/src/main/java/io/ballerina/scan/internal/StaticCodeAnalyzer.java

```diff
@@ -100,6 +116,77 @@ public void visit(ImplicitAnonymousFunctionExpressionNode implicitAnonymousFunct
         this.visitSyntaxNode(implicitAnonymousFunctionExpressionNode.expression());
     }
 
+    @Override
+    public void visit(ClassDefinitionNode classDefinitionNode) {
+        checkNonIsolatedPublicClassDefinitions(classDefinitionNode);
+        checkNonIsolatedPublicClassMembers(classDefinitionNode);
+    }
+
+    private void checkNonIsolatedPublicClassDefinitions(ClassDefinitionNode classDefinitionNode) {
+        semanticModel.symbol(classDefinitionNode).ifPresent(symbol -> {
+            if (symbol instanceof ClassSymbol classSymbol) {
+                List<Qualifier> qualifiers = classSymbol.qualifiers();
+                if (getQualifier(qualifiers, PUBLIC_KEYWORD.stringValue()) &&
+                        !getQualifier(qualifiers, ISOLATED_KEYWORD.stringValue())) {
+                    reportIssue(classDefinitionNode, CoreRule.PUBLIC_NON_ISOLATED_CONSTRUCT);
+                }
+            }
+        });
+    }
+
+    private void checkNonIsolatedPublicClassMembers(ClassDefinitionNode classDefinitionNode) {
+        classDefinitionNode.members().forEach(member -> {
+            semanticModel.symbol(member).ifPresent(symbol -> {
+                if (symbol.kind() == SymbolKind.METHOD) {
+                    checkNonIsolatedPublicMethods((FunctionDefinitionNode) member, classDefinitionNode);
+                }
+                member.accept(this);
+            });
+        });
+    }
+
+    private void checkNonIsolatedPublicMethods(FunctionDefinitionNode member,
+                                               ClassDefinitionNode classDefinitionNode) {
+        semanticModel.symbol(classDefinitionNode).ifPresent(symbol -> {
+            if (symbol instanceof ClassSymbol classSymbol) {
+                if (getQualifier(classSymbol.qualifiers(), PUBLIC_KEYWORD.stringValue()) &&
+                        getQualifier(member.qualifierList(), PUBLIC_KEYWORD) &&
+                        !getQualifier(member.qualifierList(), ISOLATED_KEYWORD)) {
+                    reportIssue(member, CoreRule.PUBLIC_NON_ISOLATED_CONSTRUCT);
+                }
+            }
+        });
+    }
+
+    private void checkNonIsolatedPublicFunctions(FunctionDefinitionNode functionDefinitionNode) {
+        semanticModel.symbol(functionDefinitionNode).ifPresent(symbol -> {
+            if (symbol.kind() != SymbolKind.METHOD) {
+                NodeList<Token> qualifiers = functionDefinitionNode.qualifierList();
+                if (getQualifier(qualifiers, PUBLIC_KEYWORD) && !getQualifier(qualifiers, ISOLATED_KEYWORD)) {
+                    reportIssue(functionDefinitionNode, CoreRule.PUBLIC_NON_ISOLATED_CONSTRUCT);
+                }
+            }
+        });
+    }
+
+    private boolean getQualifier(List<Qualifier> qualifierList, String qualifierValue) {
```

Comment:
Can't we simply do a `contains` check here. qualifierList.contains(qualifierValue)

Reply:
Qualifier is a list of enum while qualifiervalue is a SyntaxKind